tensornetwork/linalg/operations.py

```diff
@@ -0,0 +1,259 @@
+  if not all_backends_same:
+    errstr = "All Tensors fed to " + fname + "must have the same backend."
+    errstr += "Backends were: \n"
+    errstr += str([name + "\n" for name in backend_names])
+  return (all_backends_same, errstr)
+
+
+def tensordot(a: Tensor, b: Tensor, axes: Sequence[Sequence[int]]) -> Tensor:
+  """Do a tensordot (contraction) of Tensors `a` and `b` over the given axes.
+  The behaviour of this function largely matches that of np.tensordot.
+
+  Args:
+    a: A Tensor.
+    b: Another Tensor.
+    axes: Two lists of integers. These values are the contraction
+          axes.
+  Raises:
+    ValueError, if a and b have different backends.
+  Returns:
+    The result of the tensordot, a Tensor.
+  """
+  if a.backend.name != b.backend.name:
+    errstr = "Tried to Tensordot Tensors with differing backends \n"
+    errstr += a.backend.name + "and " + b.backend.name + "."
+    raise ValueError(errstr)
+  out_array = a.backend.tensordot(a.array, b.array, axes)
+  out_tensor = Tensor(out_array, backend=a.backend)
+  return out_tensor
+
+
+def reshape(tensor: Tensor, new_shape: Sequence[int]) -> Tensor:
+  """Reshape Tensor to the given shape.
+
+  Args:
+    tensor: Tensor to reshape.
+    new_shape: The new shape.
+  Returns:
+    The reshaped Tensor.
+  """
+  return tensor.reshape(new_shape)
+
+
+def transpose(tensor: Tensor, perm: Optional[Sequence[int]] = None) -> Tensor:
+  """ Return a new `Tensor` transposed according to the permutation set
+  by `axes`. By default the axes are reversed.
+  Args:
+    axes: The permutation. If None (default) the index order is reversed.
+  Returns:
+    The transposed `Tensor`.
+  """
+  return tensor.transpose(perm=perm)
+
+
+def take_slice(tensor: Tensor, start_indices: Tuple[int, ...],
+               slice_sizes: Tuple[int, ...]) -> Tensor:
+  """Obtains a slice of a Tensor based on start_indices and slice_sizes.
+
+  Args:
+    Tensor: A Tensor.
+    start_indices: Tuple of integers denoting start indices of slice.
+    slice_sizes: Tuple of integers denoting size of slice along each axis.
+  Returns:
+    The slice, a Tensor.
+  """
+  sliced = tensor.backend.slice(tensor.array, start_indices, slice_sizes)
+  sliced_tensor = Tensor(sliced, backend=tensor.backend)
+  return sliced_tensor
+
+
+def concatenate(tensors: Sequence[Tensor],
+                axis: Optional[int] = 0) -> Tensor:
+  """Concatenate a sequence of Tensors together about the given axis.
+  Args:
+    tensors: The list of Tensors to concatenate.
+    axis   : The axis to concatenate on. Default 0.
+  Returns:
+    The concatenated Tensor.
+  """
+  if len(tensors) <= 1:
+    raise ValueError("Must supply at least two Tensors to concatenate.")
+  all_backends_same, errstr = _check_backends(tensors, "concatenate")
+  if not all_backends_same:
+    raise ValueError(errstr)
+  arrays = [tensor.array for tensor in tensors]
+  backend = tensors[0].backend
+  concat_array = backend.shape_concat(arrays, axis)
```

Comment:
shape_concat doesn't do what you think it does. `shape_concat` should only be used to concatenate shapes of tensors, not tensors

Reply:
Fair enough, but in this case the relevant docstrings in abstract_backend.py are perhaps misleading

Reply:
for example, jax_backend uses numpy to concatenate. If you're running on GPU and try to concatenate you'll pull data off and to the GPU unneccessarily

Reply:
yeah, you're right. This should be fixed